File: api/download.js
// api/download.js - Simple redirect-based solution (no ytdl-core dependency)

export default function handler(req, res) {
  // Enable CORS
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

  // Handle preflight request
  if (req.method === 'OPTIONS') {
    return res.status(200).end();
  }

  // Get parameters
  const { url, type } = req.query;

  // Validate input
  if (!url) {
    return res.status(400).json({ 
      success: false,
      error: 'URL parameter is required' 
    });
  }

  try {
    if (type === 'youtube') {
      // Extract video ID from YouTube URL
      const videoIdMatch = url.match(/(?:youtube\.com\/watch\?v=|youtu\.be\/|youtube\.com\/embed\/|youtube\.com\/shorts\/)([a-zA-Z0-9_-]{11})/);
      
      if (!videoIdMatch) {
        return res.status(400).json({ 
          success: false,
          error: 'Invalid YouTube URL',
          message: 'Could not extract video ID from URL'
        });
      }

      const videoId = videoIdMatch[1];
      
      // Return redirect URLs to working download services
      return res.status(200).json({
        success: true,
        videoId: videoId,
        title: 'YouTube Video',
        thumbnail: `https://img.youtube.com/vi/${videoId}/maxresdefault.jpg`,
        downloadServices: [
          {
            name: 'SaveFrom.net',
            url: `https://en.savefrom.net/#url=https://www.youtube.com/watch?v=${videoId}`,
            description: 'Multiple quality options'
          },
          {
            name: 'Y2Mate',
            url: `https://www.y2mate.com/youtube/${videoId}`,
            description: 'HD quality downloads'
          },
          {
            name: 'SSYouTube',
            url: `https://ssyoutube.com/watch?v=${videoId}`,
            description: 'Fast downloads'
          }
        ],
        message: 'Click on any service below to download the video'
      });

    } else if (type === 'twitter') {
      return res.status(200).json({
        success: true,
        downloadServices: [
          {
            name: 'SSSTwitter',
            url: `https://ssstwitter.com/?url=${encodeURIComponent(url)}`,
            description: 'Fast and reliable'
          },
          {
            name: 'TWMate',
            url: `https://twmate.com/en2?url=${encodeURIComponent(url)}`,
            description: 'HD quality support'
          }
        ],
        message: 'Click on any service below to download the video'
      });

    } else {
      return res.status(400).json({ 
        success: false,
        error: 'Invalid type',
        message: 'Type must be either "youtube" or "twitter"'
      });
    }

  } catch (error) {
    console.error('API Error:', error);
    return res.status(500).json({ 
      success: false,
      error: 'Server error',
      message: error.message || 'An unexpected error occurred'
    });
  }
}

File: api/download.py
from http.server import BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs
import yt_dlp
import json

class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        # Enable CORS
        self.send_response(200)
        self.send_header('Content-type', 'application/json')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()

        try:
            # Parse query parameters
            query_components = parse_qs(urlparse(self.path).query)
            video_url = query_components.get('url', [None])[0]
            
            if not video_url:
                error_response = json.dumps({
                    'success': False,
                    'message': 'Missing URL parameter'
                })
                self.wfile.write(error_response.encode())
                return

            # yt-dlp options for best compatibility
            ydl_opts = {
                'format': 'bestvideo[ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best',
                'quiet': True,
                'no_warnings': True,
                'extract_flat': False,
                'nocheckcertificate': True,
                'user_agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
            }

            # Extract video information
            with yt_dlp.YoutubeDL(ydl_opts) as ydl:
                info = ydl.extract_info(video_url, download=False)
                
                # Get available formats
                formats = []
                seen_qualities = set()
                
                if 'formats' in info:
                    for f in info['formats']:
                        # Only include video+audio or video formats
                        if f.get('vcodec') != 'none' and f.get('ext') in ['mp4', 'webm']:
                            height = f.get('height', 0)
                            if height and height not in seen_qualities and height >= 360:
                                quality_label = f"{height}p"
                                formats.append({
                                    'quality': quality_label,
                                    'url': f.get('url', ''),
                                    'ext': f.get('ext', 'mp4'),
                                    'filesize': f.get('filesize', 0)
                                })
                                seen_qualities.add(height)
                
                # Sort by quality (highest first)
                formats.sort(key=lambda x: int(x['quality'].replace('p', '')), reverse=True)
                
                # Limit to top 5 formats
                formats = formats[:5]
                
                # If no formats found, add a fallback
                if not formats:
                    formats.append({
                        'quality': 'Best Available',
                        'url': info.get('url', ''),
                        'ext': 'mp4',
                        'filesize': 0
                    })

                response = {
                    'success': True,
                    'title': info.get('title', 'Unknown'),
                    'thumbnail': info.get('thumbnail', ''),
                    'duration': info.get('duration', 0),
                    'uploader': info.get('uploader', 'Unknown'),
                    'formats': formats
                }

            self.wfile.write(json.dumps(response).encode())

        except Exception as e:
            error_response = json.dumps({
                'success': False,
                'message': f'Error: {str(e)}'
            })
            self.wfile.write(error_response.encode())

    def do_OPTIONS(self):
        # Handle CORS preflight
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, OPTIONS')
        self.send_header('Access-Control-Allow-Headers', 'Content-Type')
        self.end_headers()
